File: leetcode-submissions/931_Minimum_Falling_Path_Sum.cpp
class Solution {
public:
    int n, m ;
    int t[101][101];
    int  solve(vector<vector<int>>& matrix, int row , int col){
        if(row==m-1) return matrix[row][col];
        int minSum = INT_MAX;
        if(t[row][col]!=-1) return t[row][col];
        // row+1, col-1
        if(row+1<n and col-1>=0){
            minSum = min(minSum, matrix[row][col]+solve(matrix, row+1, col-1));
        }
        // row+1, col
        if(row+1<n){
            minSum= min(minSum, matrix[row][col]+solve(matrix, row+1, col));
        }
        // row+1, col+1
        if(row+1<n and col+1<m){
            minSum = min(minSum , matrix[row][col]+solve(matrix, row+1, col+1));
        }
        return t[row][col]=minSum ;
    }
    int minFallingPathSum(vector<vector<int>>& matrix) {
        memset(t, -1, sizeof(t));
        int row = 0;
        n= matrix.size();
        m=matrix[0].size();
        int minSum = INT_MAX;
        for(int col=0;col<m;col++){
            minSum = min(minSum ,solve(matrix, row , col) );
        }
        return minSum==INT_MAX?0:minSum ;
    }
};

File: leetcode-submissions/931_Minimum_Falling_Path_Sum.ts
function minFallingPathSum(matrix: number[][]): number {
    const m = matrix.length, n = matrix[0].length;
    const dp: number[][] = Array.from({ length: m }, () => Array(n).fill(-1));
    let minAns = Number.MAX_SAFE_INTEGER;

    for (let j = 0; j < n; j++) {
        minAns = Math.min(solve(0, j, m, n, matrix, dp), minAns);
    }

    return minAns;
}

function solve(i: number, j: number, m: number, n: number, matrix: number[][], dp: number[][]): number {
    if (i < 0 || j < 0 || i >= m || j >= n) {
        return Number.MAX_SAFE_INTEGER;
    }

    if (i === m - 1) {
        return matrix[i][j];
    }

    if (dp[i][j] !== -1) {
        return dp[i][j];
    }

    const current = matrix[i][j];
    return dp[i][j] = current + Math.min(
        solve(i + 1, j - 1, m, n, matrix, dp),
        solve(i + 1, j,     m, n, matrix, dp),
        solve(i + 1, j + 1, m, n, matrix, dp)
    );
}
